### ISSUE TITLE
Add support for required / optional properties

### ISSUE THREAD

**Issue by dsanders1234** [mobxjs/serializr#73]
Currently, if the property in the schema does not exist in the json, it is simply skipped/ignored.  It would be really useful to be able to mark properties as either optional or required and give an error if the property did not match the expected behavior.

**Comment by alexggordon** [comment]
You can mostly do that when using class schemas pretty simply. 

```
Class Person {
    @observable @serializable age;
    @observable @serializable firstName = "John"
    @observable @serializable lastName = "Doe"
}

const person = deserialize({age: 20, firstName: "Alex"}, Person)

console.log(person.lastName) // results in "Doe"
```

does that solve your problem?

**Comment by dsanders1234** [comment]
This may help for optional use case (although we are not using mob-x, just the serializr).  However what we would also like is to have an error if the property does not exist to ensure that data coming from the server is correct.

**Comment by alexggordon** [comment]
I would expect something like that to be enforced outside of Serializr. That's not really in the problem domain that serializr is trying to solve. 

However, it sounds like if you're trying to validate json that json-schema might be a good solution for the problem. There are [some json-schema validators](https://github.com/epoberezkin/ajv) that are very fast now, so might be a good option.

**Comment by dsanders1234** [comment]
Yeah, I guess I see your point.  We'll try to use TypeScript decorators or something.  I don't want to have two models of my data (a typescript class and a json schema).

**Comment by alexggordon** [comment]
Sounds good!

**Comment by dietergeerts** [comment]
I would like this to be reopened, because optional properties can happen, and thus, when not comming from the BE, they shouldn't be added to the deserialized object, and now they do with an undefined value. So an optional option would be great

**Comment by alexggordon** [comment]
So this issue is something I've discussed at length because of an issue we have. There's a couple things that go into it, that have ultimately led us to create app specific serializr types. 

If you think of the possible situations in which there could be a undefined type:

1. Undefined (or key missing) from the backend. 
2. User sets it to undefined (select menu where they can deselect it). 

Then, given those to situations, you may either 

1. Not care about the key missing in the serialized json
2. Care about the key being in the serialized json (if the user sets something to undefined, then undefined is a user value you want to persist to the DB). 

The issue is that you have to allow for both of those. For awhile, i've thought that the best way to solve them is just a custom serializr

```
import { serializable, custom, SKIP } from 'serializr';

@serializable(custom(
   value => value === undefined ? SKIP : value,
   value => value
));
```

you can then distribute that by storing it as a function:

```
import { serializable, custom, SKIP } from 'serializr';

const primitiveSkipIfUndefined = custom(
   value => value === undefined ? SKIP : value,
   value => value
)

@serializable(primitiveSkipIfUndefined);
```

However, I wonder if it would help to add in some more helper functions like `SKIP` to serializr. There's a [similar discussion](https://github.com/mobxjs/serializr/issues/58#issuecomment-334613585) going on where I've been wondering if a optional null type for arrays could be of use too. That or a `strict` mode similar to mobx that enforces types. 

Anyone have thoughts on either of those?

**Comment by dietergeerts** [comment]
I have been thinking about this, and I think it would be nice to explicitly tell the schema that it's a real optional value, which means that OR the key is there OR the key isn't there, which has nothing to do with the actual value.

**Comment by alexggordon** [comment]
That's what my example does. if the value is undefined it skips the key entirely. It is omitted from the json. 

If you're referring to something different, could you give an example of what it would look like? Value based omission is the only way I could imagine an optional serializr type to work.

And while you're correct, with regards to javascript, unfortunately since this project serializes to json, and there's no difference between undefined and null in json, it's not really a super helpful distinguishment to make :(

**Comment by chengjianhua** [comment]
Is it ok to have a type called `optional` ?

In most case, the type user defined are treated as required:

```
@serializable(list()) strings;
```

In the above example, `strings` is required, it could not be `null` or `undefined`, if does, the (de)serialize failed.

If we do something like:

```
@serializable(optional(list())) strings;
``

If `strings` is `null` or `undefined`, the (de)serialize will not throw error because it's marked as optional.

**Comment by alexggordon** [comment]
So there's two issues going on so let's me careful not to intermingle them. 

First is the possible omission of keys from the serialized json, based on incoming values. 

Second is arrays failing serialization because they aren't an "array like" object. I think your suggestion @chengjianhua solves the second issue, which is more being discussed in #58, but I don't think it solves the issue brought up here. Right? (correct me if I'm wrong?)

**Comment by dietergeerts** [comment]
I think the optional type would solve both, because:

- Possible omission of keys from the serialized json, based on incoming values:
-- When optional is used, then undefined/null or missing key will all end up with not being in the json, as it's really optional, meaning it can be omitted.
-- When not optional, then the key should be there, being undefined/null, as it's not optional.

- Array serialization
-- When setting optional, if not an array, then it shouldn't be in the serialized json, as it's optional
-- When not setting optional, but the value is not array-like, then it would mean that it's undefined/null

Please correct me if I miss some use cases or logical interpretations

**Comment by chengjianhua** [comment]
@alexggordon I think my suggestion aims to solve the second question you described. 

> @dietergeerts  UPDATE: The actual bug is in the serializer. When you deserialize, the property is skipped, but when you serialize, the property will be there as undefined....

In strictly, this is a bug, it behaves inconsistently while deserializing or serializing.

IMO, the behavior while property is undefined should be consistent, after solved this, we can think about the `optional` further.

**Comment by dietergeerts** [comment]
> User sets it to undefined (select menu where they can deselect it).

Note that if you follow the specs, null means that the value is actually set to nothing, while undefined is something that happens by accident, either the property not being there, or a function that doesn't return anything. So if the user has set the value to nothing, you should use null.

**Comment by alexggordon** [comment]
I think hiding this in an optional field might be a little opaque. 

After doing some research, most JSON serializers have the concept of an "ignore if null" that only applies on serialization of an object. Essentially, if the value is undefined or null, it is removed from the serialized object. 

I think a class decorator or something to that effect would be useful, and solve this issue in a more transparent way. 

References:
[Json.net](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_NullValueHandling.htm)
[OJ, ruby](http://www.ohler.com/oj/doc/file.Options.html#label-3Aallow_blank+-5BBoolean-5D)

With regard to the array deserialize/serialize, I think that should be handled separately.

**Comment by dietergeerts** [comment]
@alexggordon , keep in mind that your first link is for C#, which doesn't have the concept of undefined values. JS does, and they both have a different meaning.

Either way, it's all about a way to define optional/required handling, because a BE will sometimes give back properties with null value and others will not return them at all if they are null in the BE system. So depending on that, you'll want a different behaviour.

If you only look at JS, then the behaviour I described before is most likely the one you want, as it complies to JS as a language. But of course, data comes from the BE, so setting the strategy will be needed, as we aren't always in charge of them ourselves.

**Comment by alexggordon** [comment]
The issues I have with an optional type is that the only feature based functionality that is changed is serialization in that keys would just be removed if the value is undefined. 

The array serialization functionality is a default and validation issue that I think should be solved by the user. If you say that something is an array, is it reasonable for the library to expect it to perform like an array?  

Based on those two things is the reason I think it would be better to expand the functionality of the key removal from the serlialized format. What if a user wants to remove a key if the value is 0? That's not something the library should do automatically. 

I'm thinking of adding an options object to the serializable decorator/function (and maybe a global one) where

```javascript
class Dog {
     @serializable({ removeKeyIf: [undefined, 0] }) age;

     @serializable(list(), 
            { removeKeyIf: puppies => !puppies || puppies.length === 0 }
     ) puppies = [];

     @serializable(object(Home),
            { removeKeyIf: home => undefined }
     ) = new Home();
}
```

there's a `removeKeyIf` option that allows for a primitive, array, or function to be passed, and if that function returns true, or value matches the primitive passed, or value is one of the values in the array, then the key is removed. I think this exposes a more explicit functionality to a user that is a little easier to understand and to customize. Thoughts?

**Comment by dietergeerts** [comment]
Though I still believe an optional type would be good with a default working that ties together with what undefined means in JS. But there should be an extra way to decide if it's optional in other cases. So a type where you can give an optional array/function like in your proposal would be nice, so if the user encounters other reasons of what optional means in a certain case, (s)he can configure it.

**Comment by alexggordon** [comment]
Along that same line, another solution I see in java is a class decorator called `@JsonIgnoreType` which just wholly ignores keys of the type passed in. That is something we could do also, if we want to start a little more simply than an options hash. 

Maybe something also like 

```
@serializrIgnoreType(undefined)
class Person {
 ...
}

@serializrIgnoreType([undefined, null])
class Person {
 ...
}
```

Upside of this is for a first pass field level property exclusion is probably a little extreme, and most people probably wouldn't use that a ton. Thoughts on either of those? If you're okay with one of those I'll see if I can get a PR up.

**Comment by chengjianhua** [comment]
@alexggordon The decorator for a class is not accurate enough, if part of the properties should be ignored with undefined or null, but another part of them shouldn't.

**Comment by alexggordon** [comment]
Can you expand on that? All that would do is remove serialized properties that have the value passed into the decorator. I'm not quite sure what you mean when you say it's not accurate enough.

**Comment by chengjianhua** [comment]
Sorry, I'll expand on this:

What I mean is that the decorator for class you demonstrated affects all the properties in a class instance,  so we can't control only part of them should be removed  because If the serialized properties have the value passed into the decorator, you'll remove all of them.

What if I want to remove only part of serialized properties that meet a certain condition but others would be kept still.

**Comment by pyrogenic** [comment]
#104 adds the `optional()` decorator that, as @alexggordon suggests, skips `undefined` values on serialization. It has no effect on deserialization, as `undefined` values are already ignored.
